build/source/engine/allocspace.f90

```diff
@@ -19,405 +19,348 @@
+ ! local variables
+ integer(i4b)                          :: iStruct       ! index of data structure 
+ character(len=256)                    :: cmessage      ! error message of downwind routine
  ! initialize errors
- err=0; message="init_model/"
- ! ensure metadata structures are deallocated
- if (associated(time_meta)) deallocate(time_meta)
- if (associated(forc_meta)) deallocate(forc_meta)
- if (associated(attr_meta)) deallocate(attr_meta)
- if (associated(type_meta)) deallocate(type_meta)
- if (associated(mpar_meta)) deallocate(mpar_meta)
- if (associated(mvar_meta)) deallocate(mvar_meta)
- if (associated(indx_meta)) deallocate(indx_meta)
- if (associated(bpar_meta)) deallocate(bpar_meta)
- if (associated(bvar_meta)) deallocate(bvar_meta)
- ! allocate metadata structures
- allocate(time_meta(maxvarTime),forc_meta(maxvarForc),attr_meta(maxvarAttr),type_meta(maxvarType),&
-          mpar_meta(maxvarMpar),mvar_meta(maxvarMvar),indx_meta(maxvarIndx),&
-          bpar_meta(maxvarBpar),bvar_meta(maxvarBvar),stat=err)
- if(err/=0)then; err=20; message=trim(message)//"problemAllocateMetadata"; return; endif
- end subroutine init_metad
+ err=0; message="initStruct/"
 
+ ! loop through data structures
+ do iStruct=1,size(structInfo)
+  ! allocate space
+  select case(trim(structInfo(iStruct)%structName))
+   case('time'); call allocGlobal(nHRU, time_meta,  timeStruct,  err, cmessage)   ! model forcing data
+   case('forc'); call allocGlobal(nHRU, forc_meta,  forcStruct,  err, cmessage)   ! model forcing data
+   case('attr'); call allocGlobal(nHRU, attr_meta,  attrStruct,  err, cmessage)   ! local attributes for each HRU
+   case('type'); call allocGlobal(nHRU, type_meta,  typeStruct,  err, cmessage)   ! local classification of soil veg etc. for each HRU  
+   case('mpar'); call allocGlobal(nHRU, mpar_meta,  mparStruct,  err, cmessage)   ! model parameters
+   case('indx'); call allocGlobal(nHRU, indx_meta,  indxStruct,  err, cmessage)   ! model variables
+   case('prog'); call allocGlobal(nHRU, prog_meta,  progStruct,  err, cmessage)   ! model prognostic (state) variables
+   case('diag'); call allocGlobal(nHRU, diag_meta,  diagStruct,  err, cmessage)   ! model diagnostic variables
+   case('flux'); call allocGlobal(nHRU, flux_meta,  fluxStruct,  err, cmessage)   ! model fluxes
+   case('deriv');call allocGlobal(nHRU, deriv_meta, derivStruct, err, cmessage)   ! model derivatives
+   case('bpar'); call allocGlobal(nHRU, bpar_meta,  bparStruct,  err, cmessage)   ! basin-average parameters
+   case('bvar'); call allocGlobal(nHRU, bvar_meta,  bvarStruct,  err, cmessage)   ! basin-average variables
+   case default; err=20; message=trim(message)//'unable to identify lookup structure'; return
+  end select
+  ! check errors
+  if(err/=0)then; err=20; message=trim(message)//trim(cmessage)//'[structure =  '//trim(structInfo(iStruct)%structName)//']'; return; endif
```

Comment:
why not use the same `handle_err()` that is used in `multi_driver()`? Would need to make it available for use here though ...